src/esn.py

```diff
@@ -209,3 +425,209 @@ def update_weights(self, xt: Array, ut: Array, yt_hat: Array,
+        x = np.zeros((N, 1)) if x_init is None else x_init.copy()
+        y = np.zeros((L, 1)) if y_init is None else y_init.copy()
+        C = np.zeros((N, N)) if C_init is None else C_init.copy()
+        loaded_input = np.zeros((N, 1))
+        # time loop
+        for t in range(T):
+            u = ut[t:t+1, :].T
+            # conceptor adaptation
+            C = C + learning_rate*( np.dot(x - np.dot(C, x), x.T) - (aperture ** (-2))*C )
+            if recall:
+                loaded_input = np.dot(self.d, x)
+            # state update (with or without feedback)
+            x = np.dot(self.w_in, u) + np.dot(self.w, x) + self.b + loaded_input
+
+            if self.feedback:
+                x += np.dot(self.w_fb, y)
+            x = np.tanh(x)
+            if recall:
+                x = np.dot(C, x)
+            # compute output
+            if self.skip_connections:
+                y = np.dot(self.w_out, np.concatenate([x, u], axis=0))
+            else:
+                y = np.dot(self.w_out, x)
+            
+            if collect_states:
+                xt.append(x)
+                yt.append(y)
+                Ct.append(C)
+
+        if collect_states:
+            xt = np.concatenate(xt, axis=1).T
+            yt = np.concatenate(yt, axis=1).T
+            return xt, yt, Ct
+        else:
+            return x, y, C
+
+    def cueing(self, ut: Array, learning_rate: float, aperture: float,
+               x_init: Optional[Array] = None, y_init: Optional[Array] = None)\
+            -> Tuple[Array, Array, Array]:
+        """
+        Forward pass for cueing, collects all reservoir states, outputs, and conceptors.
+
+        :param ut: (T, K)
+        :param x_init: (N, 1)
+        :param y_init: (L, 1)
+        :param learning_rate: float, learning rate for incremental adapatation of C
+        :param aperture: float
+        :return xt: (T, N)
+        :return yt: (T, L)
+        :return Ct: (T*N, N)
+        """
+        return self._auto(ut, learning_rate, aperture, x_init, y_init,
+                          collect_states=True, recall=False)
+    
+    def forward_cueing(self, ut: Array, learning_rate: float, aperture: float,
+                       x_init: Optional[Array] = None, y_init: Optional[Array] = None)\
+            -> Tuple[Array, Array, Array]:
+        """
+        Forward pass for cueing, only returns the last reservoir state vector and output vector, and 
+        the cue conceptor.
+
+        :param ut: (T, K)
+        :param x_init: (N, 1)
+        :param y_init: (L, 1)
+        :param learning_rate: float, learning rate for incremental adapatation of C
+        :param aperture: float
+        :return x: (N, 1)
+        :return y: (L,1)
+        :return C: (N, N), cue conceptor
+        """
+        return self._auto(ut, learning_rate, aperture, x_init, y_init,
+                          collect_states=False, recall=False)
+
+    def recall(self, ut: Array, learning_rate: float, aperture: float,
```

Comment:
same as above, maybe call this `harvest_recall`?
```suggestion
    def harvest_recall(self, ut: Array, learning_rate: float, aperture: float,
```

Reply:
Okay, I changed it.